### ISSUE TITLE
Define and implement proper C++ exception hierarchy 

### ISSUE THREAD

**Issue by Fluxie** [Rantanen/intercom#45]
We could model our exception as follows:

std::logic_error
->intercom::logic_error
- -> intercom::no_such_interface
- -> ...

std::runtime_exception
-> intercom::runtime_exception
- -> ...

This was the next task I was planning to do after getting the initial wrapper implemented. I saw that you already changed some of the exceptions to runtime_exception. However, I think that at least "No such interface" exception needs to be derived from std::logic_error. 

For supporting dynamic_cast-like construct I would prefer a try-method:  try_query_interface, for example. Especially in MS-COM calling Queryinterface and checking for possible E_NOINTERFACE is a valid method for checking if a COM object implements certain functionality which makes this feature pretty much mandatory.

**Comment by Rantanen** [comment]
Yeah, I just needed some debugging information and std::exception can't carry error message (unless you're on MS platform \:D/)

I'm fine changing some of them to logic_errors.
